### PR TITLE
Ari- Octos- Muncher

### PR DESCRIPTION
**Pull request by arrriiii**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I was able to read through the documentation and utilize postman for what attributes to display. How to handle the uri for the recipe_detail method was difficult. -- Although I did not get to css, I enjoyed working through this project.
Describe your API Wrapper.  How did you decide on the methods you created? | The documentation for Edamam required a q= for query searches and r= for a recipe id, so I created two methods for each url; The recipe_detail method to return the details of a single recipe and a recipe_list for a list of recipes with the given query.
Describe an edge case or failure case test you wrote for your API Wrapper. | I wrote a case for when the recipe_list method that would raise an error when the query DNE-- the test failed not because I didn't raise a standard error but b/c of my argument error in my recipe.rb file. I found that the failing test was because one of the parameters in from_api method was spelled incorrectly. > ingredientlines was actually ingredientLines<
Explain how VCR aids in testing an API. | VCR records the HTTP calls and uses that information to mimic an Api in the test environment. -- This way, testing is not relied on the db and limits the calls to the api.
What is the Heroku URL of your deployed application? | Unfortunately, wasn't able to deploy to Heroku
Provide a link to the Trello board you used |https://trello.com/b/3GNIhB87/muncher

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'grub#new'
+  get '/recipe', to: 'grub#index', as: 'grub_list'
+  get '/recipes/:uri', to: 'grub#show', as: 'grub'
+  get '/recipe/new', to: 'grub#new'
```
Comment (droberts-sea) [inline config/routes.rb:7]: Two things here:

- I'm not sure that `new` is a good description of what this route is doing - something non-RESTful like `search` might be a better name
- Could you use `resources` for the index and show routes?

File: app/controllers/grub_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+  private
+  def catch_api_error
+    begin
+      yield
+    rescue EdamamApiWrapper::RecipeError => error
+      flash[:status] = :failure
+      flash[:message] = "API called failed: #{error}"
+      redirect_back fallback_location: root_path
```
Comment (droberts-sea) [inline app/controllers/grub_controller.rb:25]: Here you add an error to the flash, but the code in your view to display flash errors doesn't match this format. This means that the user gets redirected to the root page with no explanation of what went wrong - not a great experience.

File: lib/edamam_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+class EdamamApiWrapper
+  class RecipeError < StandardError ; end
+
+  APP_KEY = ENV["APP_KEY"]
+  APP_ID = ENV["APP_ID"]
+  BASE_URL = "https://api.edamam.com/search?"
+  URI_BASE = "http://www.edamam.com/ontologies/edamam.owl#recipe_"
+  FROM = 1
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:8]: `BASE_URL` and `URI_BASE` are very visually similar, and it's not immediately clear how they're different. Could you rename one or both to something more distinct?

File: lib/edamam_api_wrapper.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    end
+  end
+
+  def self.recipe_detail(uri)
+
+    encoded_uri = URI.encode("#{BASE_URL}r=#{URI_BASE}#{uri}&app_id=#{APP_ID}&app_key=#{APP_KEY}")
+    response = HTTParty.get(encoded_uri)
+
+    raise StandardError.new("Could not find recipe") if response == []
+
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:28]: You do a good job of checking for errors in this file!

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | yes
Comprehension questions | yes
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | yes
Errors are reported to the user | no - see inline
API Wrapper to handle the API requests | yes - good work
Controller testing | no
Lib testing | yes - good work identifying failure cases
Search Functionality | yes
List Functionality | yes
Show individual item functionality (link to original recipe opens in new tab) | yes (link does not pop out)
**Styling** | 
Responsive layout | no
List View shows 10 items at a time/pagination | yes
The app is styled to create an attractive user interface | yes
**API Features** | 
The App attributes Edaman | yes
The VCR casettes do not contain the API key | yes
**External Resources** | 
Link to Trello Board | yes
Link to deployed app on Heroku | no
**Overall** | This is a really strong start. Your lib files and corresponding tests are very solid, and it's clear that the learning goals of this assignment around working with an API were met. However while the app mostly works, there are problems around error handling and your controller tests are missing entirely, which are important parts of our overall Rails curriculum. Please focus on these as you work on VideoStore API, and keep up the hard work.
